File: ios/Classes/PostPurchase/PostPurchaseExperienceManager.swift
import Flutter
import WebKit
import KlarnaMobileSDK

class PostPurchaseExperienceManager {
    
    static let NOT_INITIAILIZED = "PostPurchaseExperience is not initialized"
    
    static func notInitialized(result: FlutterResult?) {
        result?(FlutterError.init(code: ResultError.postPurchaseError.rawValue, message: NOT_INITIAILIZED, details: "Call 'PostPurchaseExperience.initialize' before using this"))
    }
    
    var webViewManager = WebViewManager()
    
    var navigationDelegate: PostPurchaseWKNavigationDelegate?
    
    lazy var scriptMessageHandler: PostPurchaseScriptMessageHandler = {
        let handler = PostPurchaseScriptMessageHandler()
        handler.delegate = self
        return handler
    }()
    
    var initialized = false
    
    var initResult: FlutterResult?
    var authResult: FlutterResult?
    var renderResult: FlutterResult?
    
    func initialize(method: PostPurchaseMethods.Initialize, result: @escaping FlutterResult) {
        if (webViewManager.webView != nil) {
            result(FlutterError.init(code: ResultError.postPurchaseError.rawValue, message: "Already initialized.", details: nil))
            return
        }
        
        if let hybridSDK = KlarnaHybridSDKHandler.hybridSDK {
            navigationDelegate = PostPurchaseWKNavigationDelegate(hybridSDK: hybridSDK)
        } else {
            KlarnaHybridSDKHandler.notInitialized(result: result)
            return
        }
        
        let contentController = WKUserContentController()
        contentController.add(scriptMessageHandler, name: "PPECallback")
        webViewManager.webConfiguration.userContentController = contentController
        webViewManager.initialize(result: nil)
        webViewManager.addToHybridSdk(result: nil)

        let webView = webViewManager.requireWebview()
        webView.navigationDelegate = navigationDelegate

        let url = Bundle(for: PostPurchaseHandler.self).url(forResource: "ppe", withExtension: "html")!
        if let sdkSource = method.sdkSource, !sdkSource.isEmpty {
            let html = AssetManager.readAsset(fileName: "ppe", fileExtension: "html")
            if let replaced = html?.replacingOccurrences(of: "https://x.klarnacdn.net/postpurchaseexperience/lib/v1/sdk.js", with: sdkSource) {
                webView.loadHTMLString(replaced, baseURL: nil)
            } else {
                webView.load(URLRequest.init(url: url))
            }
        } else {
            webView.load(URLRequest.init(url: url))
        }

        initResult = result
        initialized = false
        let initScript = "initialize(\(method.locale.jsScriptString()), \(method.purchaseCountry.jsScriptString()), \(method.design.jsScriptString()))"
        _ = navigationDelegate?.queueJS(webViewManager: webViewManager, script: initScript)
    }
    
    func destroy(method: PostPurchaseMethods.Destroy, result: @escaping FlutterResult) {
        webViewManager.destroy(result: nil)
        webViewManager = WebViewManager()
        navigationDelegate = nil
        initialized = false
        initResult = nil
        renderResult = nil
        authResult = nil
        result(nil)
    }
    
    func renderOperation(method: PostPurchaseMethods.RenderOperation, result: @escaping FlutterResult) {
        if (!initialized) {
            PostPurchaseExperienceManager.notInitialized(result: result)
            return
        }
        renderResult = result
        _ = navigationDelegate?.queueJS(webViewManager: webViewManager, script: "renderOperation(\(method.locale.jsScriptString()), \(method.operationToken.jsScriptString()))")
        showWebView()
    }
    
    func authorizationRequest(method: PostPurchaseMethods.AuthorizationRequest, result: @escaping FlutterResult) {
        if (!initialized) {
            PostPurchaseExperienceManager.notInitialized(result: result)
            return
        }
        authResult = result
        _ = navigationDelegate?.queueJS(webViewManager: webViewManager, script: "authorizationRequest(\(method.locale.jsScriptString()), \(method.clientId.jsScriptString()), \(method.scope.jsScriptString()), \(method.redirectUri.jsScriptString()), \(method.state.jsScriptString()), \(method.loginHint.jsScriptString()), \(method.responseType.jsScriptString()))")
        showWebView()
    }
    
    private func showWebView() {
        webViewManager.show(result: nil)
        webViewManager.webViewController?.view.backgroundColor = UIColor.clear
        webViewManager.webView?.backgroundColor = UIColor.clear
        webViewManager.webView?.isOpaque = false
    }
    
    private func hideWebView() {
        webViewManager.hide(result: nil)
    }
}

extension PostPurchaseExperienceManager: PostPurchaseScriptCallbackDelegate {
    func onInitialize(success: Bool, message: String?, error: String?) {
        if (success) {
            initResult?(message)
            initialized = true
        } else {
            initResult?(FlutterError.init(code: ResultError.postPurchaseError.rawValue, message: message, details: error))
        }
        initResult = nil
    }
    
    func onRenderOperation(success: Bool, message: String?, error: String?) {
        if (success) {
            renderResult?(message)
        } else {
            renderResult?(FlutterError.init(code: ResultError.postPurchaseError.rawValue, message: message, details: error))
        }
        renderResult = nil
        hideWebView()
    }
    
    func onAuthorizationRequest(success: Bool, message: String?, error: String?) {
        if (success) {
            authResult?(message)
        } else {
            authResult?(FlutterError.init(code: ResultError.postPurchaseError.rawValue, message: message, details: error))
        }
        authResult = nil
        hideWebView()
    }
    
    func onError(message: String?, error: Error?) {
        ErrorCallbackHandler.instance.sendValue(value: message)
        hideWebView()
    }
    
    
}

File: android/src/main/kotlin/com/klarna/inapp/sdk/flutter_klarna_inapp_sdk/postpurchase/PostPurchaseExperienceManager.kt
package com.klarna.inapp.sdk.flutter_klarna_inapp_sdk.postpurchase

import android.graphics.Color
import android.view.View
import android.webkit.WebChromeClient
import com.klarna.inapp.sdk.flutter_klarna_inapp_sdk.ErrorCallbackHandler
import com.klarna.inapp.sdk.flutter_klarna_inapp_sdk.ResultError
import com.klarna.inapp.sdk.flutter_klarna_inapp_sdk.core.util.jsScriptString
import com.klarna.inapp.sdk.flutter_klarna_inapp_sdk.core.webview.WebViewManager
import com.klarna.inapp.sdk.flutter_klarna_inapp_sdk.hybrid.KlarnaHybridSDKHandler
import io.flutter.plugin.common.MethodChannel

internal class PostPurchaseExperienceManager {

    companion object {
        private const val NOT_INITIALIZED = "PostPurchaseExperience is not initialized"

        fun notInitialized(result: MethodChannel.Result?) {
            result?.error(
                    ResultError.POST_PURCHASE_ERROR.errorCode,
                    NOT_INITIALIZED,
                    "Call 'PostPurchaseExperience.initialize' before this."
            )
        }
    }

    var webViewManager: WebViewManager = WebViewManager()

    var webViewClient: PostPurchaseExperienceWebViewClient? = null

    var initialized = false

    private var initResult: MethodChannel.Result? = null
    private var authResult: MethodChannel.Result? = null
    private var renderResult: MethodChannel.Result? = null

    private val jsInterface = PostPurchaseExperienceJSInterface(PPEResultCallback())

    fun initialize(method: PostPurchaseExperienceMethod.Initialize, result: MethodChannel.Result?) {
        if (webViewManager.webView != null || initialized) {
            result?.error(ResultError.POST_PURCHASE_ERROR.errorCode, "Already initialized.", null)
            return
        }

        val hybridSDK = KlarnaHybridSDKHandler.hybridSDK
        if (hybridSDK != null) {
            webViewClient = PostPurchaseExperienceWebViewClient(hybridSDK)
        } else {
            KlarnaHybridSDKHandler.notInitialized(result)
            return
        }

        webViewManager.initialize(null)
        webViewManager.addToHybridSdk(null)

        val webView = webViewManager.requireWebView()
        webView.webChromeClient = WebChromeClient()
        webView.webViewClient = webViewClient
        webView.settings.javaScriptEnabled = true
        webView.settings.domStorageEnabled = true
        webView.visibility = View.INVISIBLE
        webView.setBackgroundColor(Color.TRANSPARENT)

        webView.addJavascriptInterface(jsInterface, "PPECallback")
        webView.loadUrl("file:///android_asset/ppe.html")

        initResult = result
        initialized = false
        val initScript = "initialize(${method.locale.jsScriptString()}, ${method.purchaseCountry.jsScriptString()}, ${method.design.jsScriptString()})"
        webViewClient?.queueJS(webViewManager, initScript)
    }

    fun destroy(method: PostPurchaseExperienceMethod.Destroy, result: MethodChannel.Result?) {
        webViewManager.destroy(null)
        webViewManager = WebViewManager()
        webViewClient = null
        initialized = false
        initResult = null
        renderResult = null
        authResult = null
        result?.success(null)
    }

    fun renderOperation(method: PostPurchaseExperienceMethod.RenderOperation, result: MethodChannel.Result?) {
        if (!initialized) {
            notInitialized(result)
            return
        }
        renderResult = result
        webViewClient?.queueJS(webViewManager, "renderOperation(${method.locale.jsScriptString()}, ${method.operationToken.jsScriptString()})")
        showWebView()
    }

    fun authorizationRequest(method: PostPurchaseExperienceMethod.AuthorizationRequest, result: MethodChannel.Result?) {
        if (!initialized) {
            notInitialized(result)
            return
        }
        authResult = result
        webViewClient?.queueJS(webViewManager, "authorizationRequest(${method.locale.jsScriptString()}, " +
                "${method.clientId.jsScriptString()}, " +
                "${method.scope.jsScriptString()}, " +
                "${method.redirectUri.jsScriptString()}, " +
                "${method.state.jsScriptString()}, " +
                "${method.loginHint.jsScriptString()}, " +
                "${method.responseType.jsScriptString()})")
        showWebView()
    }

    private fun showWebView() {
        webViewManager.show(null)
    }

    private fun hideWebView() {
        webViewManager.hide(null)
    }

    inner class PPEResultCallback : PostPurchaseExperienceJSInterface.ResultCallback {

        override fun onInitialized(success: Boolean, message: String?, error: String?) {
            if (success) {
                initResult?.success(message)
                initialized = true
            } else {
                initResult?.error(ResultError.POST_PURCHASE_ERROR.errorCode, message, error)
            }
            initResult = null
        }

        override fun onRenderOperation(success: Boolean, message: String?, error: String?) {
            if (success) {
                renderResult?.success(message)
            } else {
                renderResult?.error(ResultError.POST_PURCHASE_ERROR.errorCode, message, error)
            }
            renderResult = null
            hideWebView()
        }

        override fun onAuthorizationRequest(success: Boolean, message: String?, error: String?) {
            if (success) {
                authResult?.success(message)
            } else {
                authResult?.error(ResultError.POST_PURCHASE_ERROR.errorCode, message, error)
            }
            authResult = null
            hideWebView()
        }

        override fun onError(message: String?, throwable: Throwable?) {
            ErrorCallbackHandler.sendValue(message)
            hideWebView()
        }
    }
}